### ISSUE TITLE
[SLPVectorizer] Assertion `Gather->ReorderIndices.empty() && "Unexpected reordering of gathers."

### ISSUE THREAD

**Issue by vedantparanjape-amd** [llvm/llvm-project#64519]
I see a crash due to SLPVectorizer with the below testcase: `opt -passes=slp-vectorizer <filename>`

```
; ModuleID = 'reduced.bc'
target datalayout = "e-m:e-p:32:32-p270:32:32-p271:32:32-p272:64:64-f64:32:64-f80:32-n8:16:32-S128"
target triple = "i386-unknown-linux-gnu"

%struct.BSS6 = type <{ [1792 x i8] }>

define internal fastcc void @test01() #0 {
  %1 = load i32, ptr getelementptr inbounds (%struct.BSS6, ptr null, i32 0, i32 0, i32 248), align 8
  %2 = load i32, ptr getelementptr inbounds (%struct.BSS6, ptr null, i32 0, i32 0, i32 252), align 4
  %3 = getelementptr double, ptr getelementptr inbounds (%struct.BSS6, ptr null, i32 0, i32 0, i32 376), i32 %2
  %4 = load double, ptr %3, align 8
  %5 = getelementptr double, ptr getelementptr inbounds (%struct.BSS6, ptr null, i32 0, i32 0, i32 376), i32 %1
  %6 = load double, ptr %5, align 8
  %7 = fadd double %6, %6
  %8 = fadd double %4, %7
  %9 = fadd double 0.000000e+00, 0.000000e+00
  %10 = fadd double %6, %9
  %11 = fadd double %10, %8
  br i1 true, label %12, label %temp_bb

temp_bb:                                      ; No predecessors!
  br label %12

12:                                               ; preds = %._crit_edge, %0
  %13 = phi double [ 0.000000e+00, %temp_bb ], [ %11, %0 ]
  ret void
}

attributes #0 = { "target-cpu"="skylake" }
```
I tried to debug the issue, it is mostly caused due to one of the nodes becoming ScatterVectorize nodes, which further adds it's operand to the tree. Here it leads to a malformed tree. I tried to do something like this and it helped. If this seems like a valid solution, I will submit a patch, thanks!

Godbolt link: https://llvm.godbolt.org/z/qnz35sv7h

```
diff --git a/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp b/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp
index b85becc4fdb7..611fb6445b45 100644
--- a/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp
+++ b/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp
@@ -4807,6 +4807,8 @@ void BoUpSLP::reorderBottomToTop(bool IgnoreReorder) {
       }
       // For gathers just need to reorder its scalars.
       for (TreeEntry *Gather : GatherOps) {
+        if (!Gather->ReorderIndices.empty())
+          reorderOrder(Gather->ReorderIndices, Mask);
         assert(Gather->ReorderIndices.empty() &&
                "Unexpected reordering of gathers.");
         if (!Gather->ReuseShuffleIndices.empty()) {
```

**Comment by vedantparanjape-amd** [comment]
CC: @alexey-bataev

**Comment by alexey-bataev** [comment]
> I see a crash due to SLPVectorizer with the below testcase: `opt -passes=slp-vectorizer <filename>`
> 
> ```
> ; ModuleID = 'reduced.bc'
> target datalayout = "e-m:e-p:32:32-p270:32:32-p271:32:32-p272:64:64-f64:32:64-f80:32-n8:16:32-S128"
> target triple = "i386-unknown-linux-gnu"
> 
> %struct.BSS6 = type <{ [1792 x i8] }>
> 
> define internal fastcc void @test01() #0 {
>   %1 = load i32, ptr getelementptr inbounds (%struct.BSS6, ptr null, i32 0, i32 0, i32 248), align 8
>   %2 = load i32, ptr getelementptr inbounds (%struct.BSS6, ptr null, i32 0, i32 0, i32 252), align 4
>   %3 = getelementptr double, ptr getelementptr inbounds (%struct.BSS6, ptr null, i32 0, i32 0, i32 376), i32 %2
>   %4 = load double, ptr %3, align 8
>   %5 = getelementptr double, ptr getelementptr inbounds (%struct.BSS6, ptr null, i32 0, i32 0, i32 376), i32 %1
>   %6 = load double, ptr %5, align 8
>   %7 = fadd double %6, %6
>   %8 = fadd double %4, %7
>   %9 = fadd double 0.000000e+00, 0.000000e+00
>   %10 = fadd double %6, %9
>   %11 = fadd double %10, %8
>   br i1 true, label %12, label %temp_bb
> 
> temp_bb:                                      ; No predecessors!
>   br label %12
> 
> 12:                                               ; preds = %._crit_edge, %0
>   %13 = phi double [ 0.000000e+00, %temp_bb ], [ %11, %0 ]
>   ret void
> }
> 
> attributes #0 = { "target-cpu"="skylake" }
> ```
> 
> I tried to debug the issue, it is mostly caused due to one of the nodes becoming ScatterVectorize nodes, which further adds it's operand to the tree. Here it leads to a malformed tree. I tried to do something like this and it helped. If this seems like a valid solution, I will submit a patch, thanks!
> 
> Godbolt link: https://llvm.godbolt.org/z/qnz35sv7h
> 
> ```
> diff --git a/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp b/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp
> index b85becc4fdb7..611fb6445b45 100644
> --- a/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp
> +++ b/llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp
> @@ -4807,6 +4807,8 @@ void BoUpSLP::reorderBottomToTop(bool IgnoreReorder) {
>        }
>        // For gathers just need to reorder its scalars.
>        for (TreeEntry *Gather : GatherOps) {
> +        if (!Gather->ReorderIndices.empty())
> +          reorderOrder(Gather->ReorderIndices, Mask);
>          assert(Gather->ReorderIndices.empty() &&
>                 "Unexpected reordering of gathers.");
>          if (!Gather->ReuseShuffleIndices.empty()) {
> ```

Thanks for the report. Your fix is not quite correct, we should not do it there. Instead, need to do an extra analysis for ScatterVectorize nodes. If they have associated ReorderIndices, need to perform same transformation, as for regular Vectorize nodes. I will fix it ASAP.

**Comment by alexey-bataev** [comment]
Fixed in d0e3a571e7f7eced43d85891181dad3fb6c0eaa1
